src/Device/TCodeDevice.cs

```diff
@@ -96,34 +138,24 @@ public void UpdateValues(IOutputTarget outputTarget)
             UpdateA0(); UpdateA1(); UpdateA2();
 
             _stringBuilder.Length = 0;
-            var l0 = AppendIfChanged(_stringBuilder, "L0", XCmd[0], ref LastXCmd[0]);
-            var l1 = AppendIfChanged(_stringBuilder, "L1", XCmd[1], ref LastXCmd[1]);
-            var l2 = AppendIfChanged(_stringBuilder, "L2", XCmd[2], ref LastXCmd[2]);
-            var r0 = AppendIfChanged(_stringBuilder, "R0", RCmd[0], ref LastRCmd[0]);
-            var r1 = AppendIfChanged(_stringBuilder, "R1", RCmd[1], ref LastRCmd[1]);
-            var r2 = AppendIfChanged(_stringBuilder, "R2", RCmd[2], ref LastRCmd[2]);
-            var v0 = AppendIfChanged(_stringBuilder, "V0", ECmd[0], ref LastECmd[0]);
-            var a0 = AppendIfChanged(_stringBuilder, "A0", ECmd[1], ref LastECmd[1]);
-            var a1 = AppendIfChanged(_stringBuilder, "A1", ECmd[2], ref LastECmd[2]);
-            var a2 = AppendIfChanged(_stringBuilder, "A2", ECmd[3], ref LastECmd[3]);
+            AppendIfChanged(_stringBuilder, "L0", XCmd[0], LastXCmd[0]);
+            AppendIfChanged(_stringBuilder, "L1", XCmd[1], LastXCmd[1]);
+            AppendIfChanged(_stringBuilder, "L2", XCmd[2], LastXCmd[2]);
+            AppendIfChanged(_stringBuilder, "R0", RCmd[0], LastRCmd[0]);
+            AppendIfChanged(_stringBuilder, "R1", RCmd[1], LastRCmd[1]);
+            AppendIfChanged(_stringBuilder, "R2", RCmd[2], LastRCmd[2]);
+            AppendIfChanged(_stringBuilder, "V0", ECmd[0], LastECmd[0]);
+            AppendIfChanged(_stringBuilder, "A0", ECmd[1], LastECmd[1]);
+            AppendIfChanged(_stringBuilder, "A1", ECmd[2], LastECmd[2]);
+            AppendIfChanged(_stringBuilder, "A2", ECmd[3], LastECmd[3]);
+
+            XCmd.CopyTo(LastXCmd, 0);
+            RCmd.CopyTo(LastRCmd, 0);
+            ECmd.CopyTo(LastECmd, 0);
 
             var data = $"{_stringBuilder}\n";
             if (!string.IsNullOrEmpty(data.Trim()))
                 outputTarget?.Write(data);
```

Comment:
`Trim()` will cause another allocation, should probably only check if `_stringBuilder.Length > 0`, then append `\n` and write to output

Reply:
could also use IsNullOrWhiteSpace if it's supported by the used runtime version

Reply:
nvm it's not part of this api version